ipatests/test_integration/test_hbac_functional.py

```diff
@@ -0,0 +1,4558 @@
+        self.run_hbactest(self.USER_1, self.master.hostname, "ftp",
+                          rule=self.HBAC_RULE,
+                          expected_patterns=[
+                              r"Access granted: True",
+                              rf"Matched rules: {self.HBAC_RULE}"
+                          ])
+        self.run_hbactest(
+            self.USER_1, self.master.hostname, "vsftpd",
+            rule="rule2",
+            expected_patterns=[
+                r"Unresolved rules in --rules",
+                r"Non-existent or invalid rules: rule2"
+            ]
+        )
+        self.run_hbactest(
+            self.USER_1, self.master.hostname, "vsftpd",
+            rule="rule2",
+            expected_patterns=[
+                r"Unresolved rules in --rules",
+                r"Non-existent or invalid rules: rule2"
+            ]
+        )
+        self.run_hbactest(
+            self.USER_2, self.master.hostname, "vsftpd",
+            rule="rule2",
+            expected_patterns=[
+                r"Unresolved rules in --rules",
+                r"Non-existent or invalid rules: rule2"
+            ]
+        )
+        self.run_hbactest(self.USER_1, self.master.hostname, "vsftpd",
+                          should_allow=True, rule=self.HBAC_RULE,
+                          nodetail=True)
+
+        # Client 1 Test
+        tasks.clear_sssd_cache(self.clients[0])
+        self.clients[0].run_command(
+            ["getent", "-s", "sss", "passwd", self.USER_1]
+        )
+        self.clients[0].run_command(["kdestroy", "-A"])
+        assert self.ftp_auth_success(
+            self.USER_1, self.USER_PASSWORD,
+            self.master, self.clients[0]
+        )
+        assert self.ftp_auth_failure(
+            self.USER_2, self.USER_PASSWORD,
+            self.master, self.clients[0]
+        )
+
+        # Client 2 Test
+        tasks.clear_sssd_cache(self.clients[1])
+        self.clients[1].run_command(
+            ["getent", "-s", "sss", "passwd", self.USER_1]
+        )
+        self.clients[1].run_command(["kdestroy", "-A"])
+        assert self.ftp_auth_success(
+            self.USER_1, self.USER_PASSWORD,
+            self.master, self.clients[1]
+        )
+
+    # Test 004: Hostgroup access (verifies bug 733663)
+    def test_hbacsvc_master_004(self, request):
+        """hbacsvc_master_004: user access to hostgroup, on master,
+        add rule"""
+        request.addfinalizer(lambda: self.cleanup_resources(
+            hbacrules=[self.HBAC_RULE],
+            hostgroups=[self.HOSTGROUP_1]
+        ))
+        tasks.kinit_admin(self.master)
+        # Create hostgroup
+        self.master.run_command(
+            ["ipa", "hostgroup-add", self.HOSTGROUP_1,
+             f"--desc={self.HOSTGROUP_1}"]
+        )
+        self.master.run_command(
+            ["ipa", "hostgroup-add-member", self.HOSTGROUP_1,
+             f"--hosts={self.clients[1].hostname}"]
+        )
+
+        # Create rule
```

Comment:
Define the rule: allow ssh access for user1 to hosts member of the hostgroup (=client2)